docs(concepts): more on component interoperability (#630)
akka/akka-sdk: docs/src/modules/concepts/pages/concepts.adoc

== Component interoperability

Akka components are able to interoperate with one another regardless of where the component may be executing. While you create systems that define the relationships between each of the components in code, at runtime instances of each component can run on different nodes. The Akka runtime provides location transparency, automating the routing of a call from one component to another even if they reside in different locations on different networks.

There are two ways to achieve interoperability between Akka components and the outside world.

> [cols="1,3", options="header"]

Author: ```suggestion
[cols="1,3", options="header"]
[.compact]
```

|===
|Client Type |Description

|xref:java:component-and-service-calls.adoc[ComponentClient]
|One component can directly invoke another component. Akka treats these direct invocations as non-blocking, asynchronous messages, meaning that the invoking component immediately returns after making the invocation and can invoke any component regardless of its location.

> For example, a Workflow component handling a long-running business transaction can invoke an Agent component to perform a task, and the Workflow will continue processing without waiting for the Agent to complete its work.

Reviewer: Maybe better to say that it can continue processing. 

By default, workflows steps are executed sequentially.

> For example, a Workflow component handling a long-running business transaction can invoke an Agent component to perform a task, and the Workflow will continue processing without waiting for the Agent to complete its work.

Author: ```suggestion
For example, a Workflow component handling a long-running business transaction can invoke an Agent component to perform a task, and the Workflow can continue processing without waiting for the Agent to complete its work.
```

|Events
|Components can emit events that can be subscribed to by other components. For example, an entity can emit an event when its state is updated. Other components can subscribe to this event, similar to how clients subscribe to Kafka topics. Events are propagated transparently between components through reliable, brokerless messaging.

Components can also subscribe to events that are incoming from external sources, which can be 3rd party brokers, APIs, or real-time streams of data.
|===

> Due to this structure, even though the code you create which connects different components together, all communications between these components are decoupled from one another with non-blocking, asynchronous communication. This enables instances of your components to run anywhere with the most efficient use of compute, memory, and I/O while the Akka runtime takes responsibility for routing requests to the location that needs to consume them. This basic behavior is one reason why Akka services, whether stateful or stateless, can scale to 10M TPS.

Author: ```suggestion
Akka is designed so that components remain loosely coupled and can communicate with each other regardless of where they run. Inter-component communication runs on virtual threads managed by the Akka runtime. These threads are lightweight and isolated in a way that ensures system resources remain available for other tasks. This allows components to interact without interfering with each other or affecting overall responsiveness. This basic behavior is one reason why Akka services, whether stateful or stateless, can scale to 10M transaction per second (TPS).
```

> Due to this structure, even though the code you create which connects different components together, all communications between these components are decoupled from one another with non-blocking, asynchronous communication. This enables instances of your components to run anywhere with the most efficient use of compute, memory, and I/O while the Akka runtime takes responsibility for routing requests to the location that needs to consume them. This basic behavior is one reason why Akka services, whether stateful or stateless, can scale to 10M TPS.

Reviewer: We recently added support for Java's Virtual Threads, and we recommend using them when calling the component client, specifically, by using the `invoke()` method instead of `invokeAsync()`.

Although `invoke()` is a blocking call, it runs on virtual threads backed by an executor optimized for blocking I/O. This setup makes blocking calls safe and efficient because virtual threads are cheap and don't tie up platform (OS) threads.

So while the operation is blocking in the traditional sense, it's not a problem, because it's not blocking a limited or heavyweight resource. It's effectively non-blocking from the application's perspective.

I'm not suggesting that we should go that technical in the concepts section, but instead of saying it's non-blocking and asynchronous all the way down, we can say that all blocking IO is done in a controlled way to avoid starvation.

Some suggestions from ai friend of mine 😄  

> Blocking operations are handled in a controlled way—using lightweight virtual threads—to ensure they don’t exhaust or block critical system resources.

or 

> Blocking work is safely isolated on virtual threads, so it doesn’t interfere with the system’s ability to stay responsive.

[cols="2,3", options="header"]
|===

> | Example Interoperability | Description

Reviewer: The use cases here doesn't seem to be connect. 

I'm referring to the `+` part. 

For me, it should read like: 

**Endpoint → Workflow → Agent**

A user triggers an HTTP request to start a long-running Workflow. The Workflow manages file processing and passes relevant data or control to an Agent, which will later use it for answering questions.

**Endpoint → Entity → View**

Another Endpoint saves user interaction history into an Entity. A View component then reads from that Entity to reconstruct the user’s conversation history for display or retrieval.

Reviewer: I see, but why are we talking about them together? It's not that they use cases a related.

Author: In the “Component Interoperability” section of the document referenced in https://github.com/lightbend/kalix/issues/13639,  I believe Tyler’s intent in discussing these components together is to illustrate that they can be deployed as a single service.
